components/feature/push/src/main/java/mozilla/components/feature/push/AutoPushFeature.kt

```diff
@@ -0,0 +1,325 @@
+import kotlinx.coroutines.plus
+import kotlinx.coroutines.runBlocking
+import mozilla.appservices.push.PushError as RustPushError
+import mozilla.appservices.push.SubscriptionResponse
+import mozilla.components.concept.push.Bus
+import mozilla.components.concept.push.EncryptedPushMessage
+import mozilla.components.concept.push.PushError
+import mozilla.components.concept.push.PushProcessor
+import mozilla.components.concept.push.PushService
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.base.observer.Observable
+import mozilla.components.support.base.observer.ObserverRegistry
+import java.io.File
+import java.util.UUID
+import java.util.concurrent.Executors
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * A implementation of a [PushProcessor] that lives as a singleton. It receives messages from a service and
+ * forwards them to be decrypted and routed. Should be created with the PushProvider.
+ *
+ * <code>
+ *     class Application {
+ *          override fun onCreate() {
+ *              val feature = AutoPushFeature(context, service, configuration)
+ *              PushProvider.install(push)
+ *          }
+ *     }
+ * </code>
+ *
+ * Subscription information for each registered [PushType] can be retrieved asynchronously:
+ *
+ * <code>
+ *    feature.registerForSubscriptions(object: Bus.Observer<PushType, String> {
+ *        override fun onSubscriptionAvailable(subscription: AutoPushSubscription) { }
+ *    })
+ *    feature.subscribeForType(PushType.Services)
+ * </code>
+ *
+ * Push messages for each registered [PushType] can also be retrieved asynchronously:
+ *
+ * <code>
+ *     feature.registerForPushMessages(object: Bus.Observer<PushType, String> {
+ *        override fun onEvent(message: String) { }
+ *     })
+ * </code>
+ *
+ */
+@Suppress("TooManyFunctions")
+class AutoPushFeature(
+    private val context: Context,
+    private val service: PushService,
+    config: PushConfig,
+    coroutineContext: CoroutineContext = Executors.newSingleThreadExecutor().asCoroutineDispatcher(),
+    private val connection: PushConnection = RustPushConnection(
+        senderId = config.senderId,
+        serverHost = config.serverHost,
+        socketProtocol = config.protocol,
+        serviceType = config.serviceType,
+        databasePath = File(context.filesDir, DB_NAME).canonicalPath
+    )
+) : PushProcessor {
+
+    private val logger = Logger("AutoPushFeature")
+    private val subscriptionObservers: Observable<PushSubscriptionObserver> = ObserverRegistry()
+    private val messageObserverBus: Bus<PushType, String> = MessageBus()
+    private val prefToken = preferences(context).getString(PREF_TOKEN, null)
+
+    internal var job: Job = SupervisorJob()
+    private val scope = CoroutineScope(coroutineContext) + job
+
+    init {
+        // If we have a token, initialize the rust component first.
+        prefToken?.let { token ->
+            runBlocking(coroutineContext) {
```

Comment:
do you actually need to block on this? Won't `scope.launch { }` work?

Reply:
If we have a token stored locally, we need to make sure the push API is init'd with that first. If we register a new token, and then the old one messages would never be delivered. This is a small cost since initialization of the Push API is just instantiation.

Reply:
See below comment. Since access to the API is synchronized, this runBlocking is no longer needed.